help_function/src/onedpl_test_malloc_free.cpp

```diff
@@ -0,0 +1,226 @@
+        dpct::malloc<int64_t>(device_exec, num_elements);
+    int64_t num_data[10];
+    std::iota(std::rbegin(num_data), std::rend(num_data), 1);
+
+    dpct::get_default_queue().submit([&](sycl::handler &h) {
+      h.memcpy(int64_ptr, &num_data, sizeof(int64_t) * num_elements);
+    });
+    dpct::get_default_queue().wait();
+
+    dpct::get_default_queue().submit([&](sycl::handler &h) {
+      h.memcpy(&num_data, int64_ptr, num_elements * sizeof(int64_t));
+    });
+    dpct::get_default_queue().wait();
+
+    std::string test_name =
+        "malloc and free int64_t array allocation - device memory with tag";
+    for (std::size_t i = 0; i != num_elements; ++i)
+      failed_tests += ASSERT_EQUAL(test_name, num_data[i],
+                                   static_cast<int64_t>(num_elements - i));
+
+    test_passed(failed_tests, test_name);
+
+    dpct::free(device_exec, int64_ptr);
+  }
+
+  // Test four, dpct::malloc and dpct::free allocation for a certain number of
+  // int64_t elements on host
+  {
+    std::size_t num_elements = 10;
+    dpct::host_sys_tag host_exec;
+    dpct::tagged_pointer<dpct::host_sys_tag, int64_t> int64_ptr =
+        dpct::malloc<int64_t>(host_exec, num_elements);
+    int64_t num_data[10];
+    std::iota(std::rbegin(num_data), std::rend(num_data), 1);
+
+    ::std::memcpy(int64_ptr, &num_data, sizeof(int64_t) * num_elements);
+
+    std::string test_name =
+        "malloc and free int64_t array allocation - host memory with tag";
+    for (std::size_t i = 0; i != num_elements; ++i)
+      failed_tests += ASSERT_EQUAL(test_name, int64_ptr[i],
+                                   static_cast<int64_t>(num_elements - i));
+
+    test_passed(failed_tests, test_name);
+
+    dpct::free(host_exec, int64_ptr);
+  }
+
+  // Test five, dpct::malloc and dpct::free allocation for a certain number of
+  // int64_t elements on device. oneDPL device policy passed as location
+  {
+    std::size_t num_elements = 10;
+    auto policy = oneapi::dpl::execution::dpcpp_default;
+    dpct::tagged_pointer<dpct::device_sys_tag, int64_t> int64_ptr =
+        dpct::malloc<int64_t>(policy, num_elements);
+    int64_t num_data[10];
+    std::iota(std::rbegin(num_data), std::rend(num_data), 1);
+
+    dpct::get_default_queue().submit([&](sycl::handler &h) {
+      h.memcpy(int64_ptr, &num_data, sizeof(int64_t) * num_elements);
+    });
+    dpct::get_default_queue().wait();
+
+    dpct::get_default_queue().submit([&](sycl::handler &h) {
+      h.memcpy(&num_data, int64_ptr, num_elements * sizeof(int64_t));
+    });
+    dpct::get_default_queue().wait();
+
+    std::string test_name =
+        "malloc and free int64_t array allocation - device memory with policy";
+    for (std::size_t i = 0; i != num_elements; ++i)
+      failed_tests += ASSERT_EQUAL(test_name, num_data[i],
+                                   static_cast<int64_t>(num_elements - i));
+
+    test_passed(failed_tests, test_name);
+
+    dpct::free(policy, int64_ptr);
+  }
```

Comment:
May not be worth the effort, but this same structure is repeated a number of times and could perhaps be unified with a test harness and a few template arguments.

Reply:
I have simplified these tests into a test function for host allocations and a test function for device allocations.